lib/index.d.ts

```diff
@@ -0,0 +1,248 @@
+import { IDisposable } from 'phosphor-disposable';
+import { Message } from 'phosphor-messaging';
+import { IChangedArgs, Property } from 'phosphor-properties';
+import { ISignal, Signal } from 'phosphor-signaling';
+import { Widget, ResizeMessage } from 'phosphor-widget';
+/**
+ * An interface required for implementing the editor model
+ */
+export interface IEditorModel {
+    /**
+     * Updates the buffer stored in the model.
+     *
+     * @param value - A string containing the complete data in the view.
+     */
+    updateBuffer(value: string): void;
+    /**
+     * Saves the current data in the existing file.
+     *
+     * @param data - A string containing the complete data to be saved.
+     */
+    save(data: string): void;
+    /**
+     * Renames the current file.
+     *
+     * @param name - A string denoting the new filename.
+     *
+     * #### Notes
+     * This does *not* remove the existing file, it simply saves a
+     * new file with the same data.
+     */
+    rename(name: string): void;
+    /**
+     * A signal emitted when the editor model state changes.
+     */
+    changed: ISignal<EditorModel, IChangedArgs<any>>;
+    /**
+     * The mode of the editor, eg. for code this would be
+     * the language, like 'python' or 'javascript'.
+     */
+    mode: string;
```

Comment:
should we use a mimetype here? And then map from that mime to a codemirror mode inside the widget.